documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+, and assuming we are going through the expression tree in a postfix way, we do not need to keep track of where to emit sub-expressions, they already were.
+Cons: Will generate many loops and temporaries that we know without deep analysis are not needed.
+
+#### One "Perfectly" Elemental Expression Sub-Tree - One Loop
+`x + y + z` is lowered into a single loop that directly computes `x(i) + y(i) + z(i)` for each element and store it into a single temp.
+When one of the array sub-expression is not elemental, it is computed on its own outside of the loop and stored into another temp that is then accessed before the loop.
+For instance, `x + foo(y) + z` where `foo(y)` is an array ends up:
+- compute `foo(y)` into `tmp1`
+- compute `x + tmp1 + z`
+With this alternative, transformational such as `transpose` are still treated as `foo`.
+Pros: Generates less FIR, the original expression structure is better preserved inside the loop.
+Cons: Slightly more complex lowering, we need to keep track of the loop where we want to emit the "inlined" operations and where to insert
+the computation of sub-expressions that are not "inlined".
+
+#### One "Affine" Elemental Expression Sub-Tree - One Loop
+Same as before, but also apply transformationals as index re-mapping instead of computing it outside of the loop.
+For instance `x + transpose(y) + z` becomes a loop computing each `x(i, j) + y(j, i) + z(i, j)`.
+Pros: Generates minimal FIR. Not much more complex that the previous one (mostly the same thing).
+Cons: Requires more knowledge about transformationals, an index mapping state must be propagated while lowering.
+
+
+## Examples
+Below are illustrations of Fortran to FIR lowering based on the above design (with one node, one loop). 
+
+### Simple examples
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x
+```
+Here, `x` expression should not be split into a loop.
+
+FIR:
+```
+ %1 = fir.load %x // that is the whole expression evaluation
+ // This is actually the assignment lowering
+ %fir.store %1 to %y
+```
+Question: This is more related to assignment lowering, but what will decide whether `%1`
+is has to be copied inside a temporary or not (e.g. in case of pointers that may overlap or not).
+Is it implicit in here because `%1` requires a storage later and
+FIR will determine whether `%y` can directly be used ? 
+### Outline scalar sub-expression outside of the loop
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x + cos(1.)
+```
+FIR:
+```
+  %tmp = fir.alloca !fir.array<4:f32>
+  %cst = constant 1.000000e+00 : f32
+  %1 = fir.call "cos" %cst
+  fir.loop %i = 1 to 4 {
+    %2 = fir.extract_value %x, %i
+    %3 = fir.addf %2, %1
+    %4 = fir.coordinate_of %tmp, %i
+    fir.store %3 to %4 
+  }
+  %5 = fir.load %tmp // This is the expression result
+  // The assignment part is not part of the expression lowering
+  fir.store %5 to %y
```

Comment:
I'm wondering if we wouldn't rather expand the copy to a `fir.loop` as well. Also, why do we need a temp array here?

Reply:
Same as above, because I am assuming expression lowering is not taking storage optimizations that require aliasing analysis (`x = transpose(x)`  we need a temp, `x = x +1` we can optimize it out). I agree that in the trivial case where the symbol on the left does not appear on the right and there are no pointers we could get probably never emit this. But if we are doing storage optimizations later anyway, is there a point to do this first analysis before lowering ?